src/server/models/db/migrations/20230302210317_add_view_FirstPublishedOnList/migration.sql

```diff
@@ -0,0 +1,8 @@
+-- Create FirstPublishedOnList view with the columns, listId, firstPublished, eventId
+-- select distinct on (l.id), similar to group by, ensures only 1 entry appears per listId
+-- min(e.time) selects the earliest time a published event occurred (firstPublished)
+create view "FirstPublishedOnList" as (
+  select distinct on (l.id) "listId", min(e.time) over (partition by l.id) "firstPublished", e.id "eventId"
+  from (select time, type, id, "listItemId" from "Event" where type = 'PUBLISHED') e
```

Comment:
see additional comment above.  You could also add List.annualReviewStartDate to this view and then you can just make one DB call to find the lists that need it populated:

```
select l."id" "listId", min(e."time") "firstPublished", l."nextAnnualReviewStartDate" 
from "List" l 
inner join "ListItem" li on l."id" = li."listId" 
inner join (
    select min(e1."time") over (partition by e1."listItemId") "time", e1."listItemId"
    from "Event" e1
    where e1."type" = 'PUBLISHED'
    group by e1."time", e1."listItemId"
) e on e."listItemId" = li."id"
group by l."id", l."nextAnnualReviewStartDate"
order by l."id";
```

Reply:
I was using group by to begin with, however I wanted to see the event id (for cross referencing purposes).

You can't include a column unless used in the groupby or aggregate. but I don't want to group by or aggregate by list.id and event.id, which is why I use distinct. 

Agree on smallest to largest for the most part, but they're inner joins. afaik inner join order doesn't matter, selecting the columns you need makes more impact (instead of joining all cols)

Reply:
for this service it's unlikely to make an impact.  On more complex queries on larger datasets, the RDBMS query optimiser could get a bit squiffy and then the join order could matter but very edge casey.